Enhance: remove minuscule text from under profile icons

SlySven opened the pull request:
This finally enables us to do away with the text under the profile icons as it replicates the normal `QListWidget` behaviour to move the selection to each item which begins with the first letter of the key pressed.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: remove minuscule text from under profile icons

This finally enables us to do away with the text under the profile icons
as it replicates the normal `QListWidget` behaviour to move the selection
to each item which begins with the first letter of the key pressed.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (commented):
This does work, but not quite as good as the Qt version - in there if I have multiple profiles starting with a similar name, I can type the name of the profile and it'll find it. Here, only the last letter is considered, and while I can press the letter many times - it's not as intuitive as entering in a few first letters of the profile.

Do you think you could have this match the behaviour?

SlySven commented:
> This does work, but not quite as good as the Qt version - in there if I have multiple profiles starting with a similar name, I can type the name of the profile and it'll find it.

Oh, I thought it only considered the most recent keypress.

Gonna have to think about that...

SlySven committed:
Revise: take a different approach and enable whole name to be typed

This also makes the non-matching profiles get disabled which I think makes
the action more obvious - and gives further instructions on how to reset
things...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -2337,3 +2342,138 @@ QIcon dlgConnectionProfiles::customIcon(const QString& text) const
+    notificationAreaIconLabelError->hide();
+    notificationAreaIconLabelInformation->show();
+    notificationAreaMessageBox->show();
+    notificationAreaMessageBox->setText(tr("As you type more letters in the name of a profile all the non-matching ones "
```
Let's not show this message, it was not necessary for the last ten years - this is definitely way too much "nagging"

SlySven replied:
Yeah, but since I am actually disabling non-matching profiles we ought to let the end user know how to get them back again. I suppose I could try something more fancy that would only show this message *once* per Mudlet run - but I do not really want to add too much more :spaghetti: code here.

Also - I cannot view streamable.com URLs - they must be (like GitHub) using features that are only compatible with the most bleeding edge browsers and unlike GH (https://github.com/JustOff/github-wc-polyfill) I do not have a fix for them.

SlySven replied:
Otherwise - is this any good? :worried:

vadi2 replied:
The message does not need to be shown at all. Good UX does not require explanation!

vadi2 reviewed (commented):
Thanks a lot for continuing to investigate this :+1: 

https://streamable.com/h8y6x2

See one example where I try to find the Materia Magica profile by typing `mate` repeatedly. It works with Qt, but with our own implementation it keeps finding ateraan instead.

SlySven commented:
> See one example where I try to find the Materia Magica profile by typing `mate` repeatedly. It works with Qt, but with our own implementation it keeps finding `ateraan` instead.

My best guess is that the `m` is ending your previous attempt to match something so the `a` is the first letter of the actual search - if you press `<Esc>` as suggested to clear any stored keypresses you should find it works as expected.

vadi2 commented:
Sent a video on Discord that shows the differences, have a look

SlySven committed:
Merge branch 'development' into Enhance_removeMinusculeTextFromUnderProfileIcons

Needed to include fix for crashing on exit in TTimerUnit destructor.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
You should maybe put a timer of 2-3 seconds of no keys pressed before starting a new name circle

SlySven commented:
Actually I did dissect that video on Discord and I eventually worked out that what I suspected was, indeed,  the case:
>  My best guess is that the m is ending your previous attempt to match something so the a is the first letter of the actual search - if you press <Esc> as suggested to clear any stored keypresses you should find it works as expected.

I really don't want to have to use a timer - as it is I think the visual feed back from all the non-matching profiles becoming disabled (with the message explaining this - but which disappears as soon as the search is cleared by the first letter that does not leave *any* matching profiles) is a viable scheme as it is...

vadi2 commented:
The idea was to keep the same behaviour, not change it, because that won't be intuitive and frustrate muscle memory, though.

Kebap commented:
I did not make up the timer myself: It is instead common best practice in such list selection widgets. Try your file explorer or web browser for good measure. Try the Mudlet map areas selection drop down menu as well. - You can't restart the selection by keep on typing. You need to pause for a moment to have it restart.

SlySven commented:
> You can't restart the selection by keep on typing. You need to pause for a moment to have it restart.

I am not sure I like the idea of using timer dependent code simply because that will need a determination of the time period to use and *that* is going to be horribly subjective IMVHO - how about if the end user is using assistive technology (or even a screen reader). I feel the enabling/disabling idea is attractive because it is very much more visible - for me the selected item is not that well differentiated under the current release codebase - but with only a few (or ultimately only one) left enabled and not greyed out - it is hard to miss the profile(s) that are possible. I will have a closer look to see exactly how `QComboBox`s behave and see if that can be reproduced - as well as the current `QListWidget` behaviour ***with*** text. The latter wasn't as obvious as I believed - I will also need to see how it responded to `<BackSpace>`/`<Del>` to see how typos are handled and if possible corrections.

SlySven commented:
:information_source: I have just tested the behaviour of the `development` branch code to see how its `QListWidget` with text entries behave and I have found that ***it only uses the last keypress and switches in order through all the profiles that* begin *with the key pressed***! Which is how the original commit https://github.com/Mudlet/Mudlet/pull/4712/commits/db80c467b75dca94f0740a7801656be7630e9ec6 behaved - **not** as you suggested @vadi2 that one could type in the whole name and it would select the particular profile.

So how *does* everyone expect/want this process to work?

vadi2 commented:
How it's done in Qt, that was the original goal to keep that behaviour. With Qt I can type `mate` and get the `Materia Magica` profile - that doesn't seem like what you describe above, so are you sure it works that way?

SlySven commented:
> How it's done in Qt, that was the original goal to keep that behaviour. With Qt I can type mate and get the Materia Magica profile - that doesn't seem like what you describe above, so are you sure it works that way?

That wasn't what I was getting with a build of the `development` branch - let me do a clean build (and reset my setup to use the system's `qtkeychain` library so I can still build it with qmake)...

... and yeah - the `development` code does only consider the last key I press - I cannot spell out the profile name like you say you have it Vadim.

SlySven commented:
AH! With @Faenriis help I've worked it out - if you type a sequence of keys *quickly* enough that it does string them together to find a match - however if you do it slower it will reset after a fraction of a second and, as I experienced, jump around the profiles. Maybe @Kebap was right when he talked about a timer that reset things - I was thinking of over a second but in fact it is only a fraction of a second.

Okay with that knowledge I can have another think.

SlySven committed:
Revise: include a timeout on concatenating keypresses

This is to provide the same functionality that the QListWidget does for
keypresses, if the keys are typed fast enough they are concatenated to form
the text to search for. However if the user types slower than this then
the individual keys are only considered on their own. This reflects what
both Vadim and I experienced - in that he was typing fast enough to spell
out a particular profile but I wasn't and so the profile that I was having
selected was changing on each keypress.

The time between keypresses allowed so that they can be concatenated I have
set to be 1 second. This is significantly slower that the Qt code but it
is long enough to permit the temporary disablement of non-matching profiles
to be a useful visual feedback - and besides, it makes it easier to specify
one of a number of profiles with a common initial text - the longer the
timeout the more keys the user can string together!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -282,6 +279,16 @@ dlgConnectionProfiles::dlgConnectionProfiles(QWidget* parent)
         mCustomIconColors.append(QColor::fromHsv(i, 128, 255));
     }
+
+    mSearchTextTimer.setInterval(1000);
```
For easier reading!

```suggestion
    mSearchTextTimer.setInterval(1s);
```

At the top, add:

```
#include <chrono>
using namespace std::chrono_literals;
```

SlySven replied:
I'm getting *deja vu* all over again about this (`std` time intervals) - I'm not a fan of them but I guess i can do it... :slightly_frowning_face:

vadi2 commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -2337,3 +2328,162 @@ QIcon dlgConnectionProfiles::customIcon(const QString& text) const
+void dlgConnectionProfiles::slot_searchTimerTimeOut()
+{
+    reenableAllProfileItems();
+    if (notificationAreaMessageBox->text() == tr("As you type more letters in the name of a profile all the non-matching ones "
```
It's intuitive enough to work with that it needs no explanation - and it's also impossible to read it fast enough :)

SlySven replied:
Guess so - I must confess that is one incentive to drag the time interval out... :smiling_imp: ... but I'll leave the refactoring into `clearNotificationArea()` as that is still useful I think.

vadi2 replied:
Yes agree, `clearNotificationArea()` is good!

Kebap replied:
clearNotificationArea ist super good - do more of those! 🌟

vadi2 reviewed (approved):
I honestly think it's fantastic, and the highlighting makes it even better than the original Qt implementation!

Just the notification should go, it's not all too helpful.

SlySven committed:
Revise: adopt peer review suggestions

Remove the help text shown during typing the keys to select a profile - the
time it is shown for is not enough for it to be that useful.

Add `std::chrono` so we can give a time delay as `1s` instead of `1000` mS.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>
